### PR TITLE
IGNITE-21571 Get rid of exeption in EventListener#notify

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21571

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java [hunk 1]
```diff
@@ -111,12 +108,6 @@ public void parallelTest() {
         assertThat(fireFuture, willCompleteSuccessfully());
     }
 
-    private static <T extends EventParameters> EventListener<T> createEventListener(
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/event/EventProducerTest.java:114]: oh god

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 2]
```diff
@@ -1378,17 +1378,13 @@ public Set<NodeWithAttributes> logicalTopology() {
     }
 
     private void registerCatalogEventListenersOnStartManagerBusy() {
-        catalogManager.listen(ZONE_CREATE, (parameters, exception) -> inBusyLock(busyLock, () -> {
-            assert exception == null : parameters;
-
+        catalogManager.listen(ZONE_CREATE, parameters -> inBusyLock(busyLock, () -> {
```
Comment (sashapolo) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:1381]: Same thing: we can use an explicit parameter type and make this code even shorter

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java [hunk 3]
```diff
@@ -189,27 +188,15 @@ public void recover(long recoveryRevision) {
     }
 
     private void addListeners(CatalogService catalogService, MetaStorageManager metaStorageManager, IndexBuilder indexBuilder) {
-        catalogService.listen(CatalogEvent.INDEX_BUILDING, (parameters, exception) -> {
-            if (exception != null) {
-                return failedFuture(exception);
-            }
-
+        catalogService.listen(CatalogEvent.INDEX_BUILDING, parameters -> {
```
Comment (sashapolo) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:191]: Same stuff about explicit types

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java [hunk 4]
```diff
@@ -123,27 +122,15 @@ public void close() {
     }
 
     private void addListeners() {
-        catalogService.listen(CatalogEvent.INDEX_BUILDING, (parameters, exception) -> {
-            if (exception != null) {
-                return failedFuture(exception);
-            }
-
+        catalogService.listen(CatalogEvent.INDEX_BUILDING, parameters -> {
```
Comment (sashapolo) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:125]: And here

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java [hunk 5]
```diff
@@ -140,13 +140,7 @@ public CompletableFuture<Void> start() {
...
         startIndexes();
 
-        catalogService.listen(INDEX_CREATE, (parameters, exception) -> {
-            if (exception != null) {
-                return failedFuture(exception);
-            }
-
-            return onIndexCreate((CreateIndexEventParameters) parameters);
-        });
+        catalogService.listen(INDEX_CREATE, parameters -> onIndexCreate((CreateIndexEventParameters) parameters));
```
Comment (sashapolo) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java:143]: And here

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 6]
```diff
@@ -2218,17 +2216,13 @@ void testAvailableIndexEvent() {
...
 
-        manager.listen(CatalogEvent.INDEX_AVAILABLE, (parameters, exception) -> {
-            if (exception != null) {
-                fireEventFuture.completeExceptionally(exception);
-            } else {
-                try {
-                    assertEquals(indexId, ((MakeIndexAvailableEventParameters) parameters).indexId());
+        manager.listen(CatalogEvent.INDEX_AVAILABLE, parameters -> {
+            try {
+                assertEquals(indexId, ((MakeIndexAvailableEventParameters) parameters).indexId());
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2221]: why not

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 7]
```diff
@@ -2248,15 +2242,11 @@ void testAvailableIndexEvent() {
...
-                fireEventFuture.completeExceptionally(exception);
-            } else {
-                try {
-                    fireEventFuture.complete(((MakeIndexAvailableEventParameters) parameters).indexId());
-                } catch (Throwable t) {
-                    fireEventFuture.completeExceptionally(t);
-                }
+        manager.listen(CatalogEvent.INDEX_AVAILABLE, parameters -> {
+            try {
+                fireEventFuture.complete(((MakeIndexAvailableEventParameters) parameters).indexId());
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2247]: why not

File: modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java [hunk 8]
```diff
@@ -87,7 +87,7 @@ protected CompletableFuture<Void> fireEvent(T evt, P params, @Nullable Throwable
         for (int i = 0; i < listeners.size(); i++) {
             EventListener<P> listener = listeners.get(i);
 
-            CompletableFuture<Boolean> future = listener.notify(params, err);
+            CompletableFuture<Boolean> future = listener.notify(params);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java:90]: fix it

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java (hunk 2 above)
Comment (tkalkirill) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:1381]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java (hunk 3 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:191]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java (hunk 4 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java:125]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java (hunk 5 above)
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java:143]: fix it

**Review by sashapolo — APPROVED** [review]

**Pull request merged**